js/api.ts

```diff
@@ -0,0 +1,49 @@
+import { fetch, reload } from "./browser";
+import { useEffect, useState } from "react";
+import { z } from "zod";
+
+export type ApiResult<Result> =
+  | { status: "loading" }
+  | { status: "ok"; result: Result }
+  | { status: "error"; error?: unknown };
+
+export const useApiResult = <RawData, Data>({
+  RawData,
+  url,
+  parser,
+}: {
+  RawData: z.ZodType<RawData>;
+  url: string;
+  parser: (data: RawData) => Data;
+}): ApiResult<Data> => {
+  const [result, setResult] = useState<ApiResult<Data>>({ status: "loading" });
+
+  useEffect(() => {
+    fetch(url)
+      .then((response) => {
+        if (response.status === 200) {
+          return response.json();
+        } else if (response.status === 401) {
+          reload();
+
+          throw new Error("Unauthenticated");
+        } else {
+          throw new Error(`Unrecognized response code: ${response.status}`);
+        }
+      })
+      .then((json) => {
+        const jsonObject = z.object({ data: z.unknown() }).parse(json);
+
+        const data = RawData.parse(jsonObject.data);
+
+        const parsedData = parser(data);
+
+        setResult({ status: "ok", result: parsedData });
+      })
+      .catch((e: unknown) => {
+        setResult({ status: "error", error: e });
```

Comment:
Could we `console.error` this just so we see it as well?

Reply:
I actually think sending an error to Sentry would be more useful, so I added a note to [the ticket](https://app.asana.com/0/616151179860796/1207620537833336/f) for frontend Sentry support. Does that sounds good to you?